### PR TITLE
[#2030] Feature: share download and upload files

### PR DESCRIPTION
**Pull request by gbonaspetti**
#2030

## Checkpoints

The goal of this section is to help code reviewer to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

- [x] Code is clear enough
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

### CODE REVIEW

File: frontend_lib/src/component/ShareLink/ShareLink.jsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+class ShareLink extends React.Component {
+  handleCopyToClipboard = () => {
+    let tmp = document.createElement('textarea')
+    document.body.appendChild(tmp)
+    tmp.value = this.props.link
+    tmp.select()
+    document.execCommand('copy')
+    document.body.removeChild(tmp)
+    /* or https://stackoverflow.com/questions/36639681/how-to-copy-text-from-a-div-to-clipboard See what's best */
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareLink/ShareLink.jsx]: it the right way to link to the source of a non intuitive snippet copied from the web.
But you must use the comment convention we use, which would be, for you:
`// INFO - GB - 2019-07-17 - the comment`

File: frontend_lib/src/component/ShareLink/ShareLink.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    const { props } = this
+
+    return (
+      <div className='shareLink d-flex'>
+        <div className='shareLink__icon'>
+          <ComposedIcon
+            icon='link'
+            smallIcon='lock'
+            style={{color: '#ababab'}}
+            smallIconStyle={{color: '#252525'}} // See how to use Variable's variables
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareLink/ShareLink.jsx]: add the comment convention and add the file's extension of Variable.styl to make it easier to understand

File: frontend/src/component/GuestPage/ProgressBarMultipleFiles.jsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import React from 'react'
+
+require('./ProgressBarMultipleFiles.styl')
+
+export const ProgressBarMultipleFiles = props => {
```
[the remark this answers is not in the document]
Comment (gbonaspetti) [reply frontend/src/component/GuestPage/ProgressBarMultipleFiles.jsx]: I agree with the idea to reuse the component, but I'm not sure to use the same name as the progress bar component in the app file.

File: frontend_lib/src/component/PopinFixed/PopinFixedRightPart.jsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  render () {
+    const { props } = this
+
+    return (
+          <div className={classnames('wsContentGeneric__content__right', `${props.customClass}__content__right`, 'd-flex')}>
+            <div
+              className={classnames(`${props.customClass}__header`, 'wsContentGeneric__content__right__header nav')} >
+              {props.menuItemList.map(menuItem =>
+                <button
+                  className={classnames('wsContentGeneric__content__right__header__icon nav-item iconBtn m-0', {'active': menuItem.id === this.state.currentChildren})}
```
[the remark this answers is not in the document]
Comment (gbonaspetti) [reply frontend_lib/src/component/PopinFixed/PopinFixedRightPart.jsx]: the goal here is to overwrite the margin rule of iconBtn, it seemed better than redoing all the properties of this class or using !important

File: frontend_lib/src/component/ShareDownload/NewShareDownload.jsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+  handleRandomPassword = () => {
+    this.props.onChangePassword({target: {value: generateRandomPassword()}})
+    const passwordInput = document.getElementsByClassName('shareDownload__password__input')[0]
+    if (passwordInput.type === 'password') {
+      this.handleSeePassword()
+    }
+  }
+
+  handleSeePasswordInput = () => {
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareDownload/NewShareDownload.jsx]: this function should be called handleClickSeePasswordInput
Also, it seems to be a toggler but there is no button to go back to the previous state.
Maybe this was missing in the spec.

File: frontend_lib/src/component/ShareDownload/NewShareDownload.jsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    return (
+      <div className='shareDownload__new'>
+        <div className='shareDownload__title'>
+          {props.t('New share')}
+        </div>
+
+        <div className='shareDownload__email'>
+          <textarea
+            type='text'
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareDownload/NewShareDownload.jsx]: textarea tag doesn't have a type attribute
I guess this one has been forgotten during the transition from an input to a textarea

File: frontend_lib/src/component/ShareDownload/NewShareDownload.jsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+          >
+            <i className='fa fa-fw fa-question-circle' />
+          </button>
+          <Popover placement='bottom' isOpen={state.popoverOpen} target='popoverMultipleEmails' toggle={this.popoverToggle}>
+            <PopoverBody>{props.t('To add multiple recipients, separate the email addresses with a comma or space.')}</PopoverBody>
+          </Popover>
+        </div>
+
+        {state.passwordActive
+        ? <div className='shareDownload__password'>
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareDownload/NewShareDownload.jsx]: this is not quite readable.
Use parenthesis to always have your opening tags at the same indent level of the closing one

File: frontend_lib/src/component/ShareDownload/NewShareDownload.jsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            className='shareDownload__password__icon'
+            key='randomSharePassword'
+            title={props.t('Generate random password')}
+            style={{':hover': {color: props.hexcolor}}}
+            onClick={this.handleRandomPassword}
+          >
+            <i className='fa fa-fw fa-repeat' />
+          </button>
+        </div>
+        : <div className='shareDownload__password'>
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareDownload/NewShareDownload.jsx]: use parenthesis to always have the opening tag at the same indent level of the closing one

File: frontend_lib/src/component/ShareDownload/NewShareDownload.jsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+          <span className='shareDownload__password__link' onClick={this.handleSeePasswordInput}>
+            {props.t('Protect by password')}
+          </span>
+        </div>
+        }
+        <div className='d-flex mt-3'>
+          <button
+            className='shareDownload__cancel btn outlineTextBtn'
+            key='cancelNewShare'
+            onClick={props.onClickReturnToManagement}
```
Comment (Skylsmoi) [inline frontend_lib/src/component/ShareDownload/NewShareDownload.jsx]: the handle should be called handleClickCancelButton or handleClickCancelNewShare

File: frontend_lib/src/component/PopinFixed/PopinFixedRightPart.jsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import React from 'react'
+import classnames from 'classnames'
+import PropTypes from 'prop-types'
+
+class PopinFixedRightPart extends React.Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      currentChildren: props.menuItemList[0].id
```
Comment (Skylsmoi) [inline frontend_lib/src/component/PopinFixed/PopinFixedRightPart.jsx]: this isn't very safe. If props.menuItemList.length === 0, this code will fail.
I understand that PopinFixedRightPart should not be called with an empty props menuItemList but, in that case, you should specify it using propTypes.

Check what I did for frontend_lib/src/PopinFixed/PopinFixed.jsx
As usual, feel free to ask me for more info

File: frontend/src/component/Workspace/ShareFolder.jsx [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+            <div className='folder__header__triangleborder__triangle primaryColorFontLighten' />
+          </div>
+
+          <div
+            className='folder__header__dragPreview'
+            ref={props.connectDragPreview}
+          >
+            <div className='folder__header__icon'
+              title={props.t('Folder')}
+              style={{ color: props.contentType.find(c => c.slug === 'folder').hexcolor }}
```
Comment (basilelegal) [inline frontend/src/component/Workspace/ShareFolder.jsx]: what happen if you don't find an element ?

File: frontend/src/component/Workspace/ShareFolder.jsx [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+  render () {
+    const { props } = this
+
+    const folderContentList = props.workspaceContentList
+      .filter(c => c.parentId === props.folderData.id)
+      .sort((a, b) => {
+        if (a.created > b.created) {
+          return -1
+        } else {
```
Comment (basilelegal) [inline frontend/src/component/Workspace/ShareFolder.jsx]: useless else

File: frontend/src/component/Workspace/ShareFolder.jsx [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+import classnames from 'classnames'
+import ContentItem from './ContentItem.jsx'
+import {
+  PAGE
+} from '../../helper.js'
+
+require('./Folder.styl')
+
+class ShareFolder extends React.Component {
+  calculateIcon = () => {
```
Comment (basilelegal) [inline frontend/src/component/Workspace/ShareFolder.jsx]: getIcon

File: frontend/src/component/Workspace/ShareFolder.jsx [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+          <div
+            className='folder__header__dragPreview'
+            ref={props.connectDragPreview}
+          >
+            <div className='folder__header__icon'
+              title={props.t('Folder')}
+              style={{ color: props.contentType.find(c => c.slug === 'folder').hexcolor }}
+            >
+              <i className={classnames('fa fa-fw', this.calculateIcon())} />
```
Comment (basilelegal) [inline frontend/src/component/Workspace/ShareFolder.jsx]: getIcon

File: frontend/src/component/Workspace/ShareFolder.jsx [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+              )}
+            </div>
+          </div>
+
+          <div className='folder__header__status' />
+
+        </div>
+
+        <div className='folder__content'>
+          {folderContentList.map((content, i) => content.type === 'folder'
```
Comment (basilelegal) [inline frontend/src/component/Workspace/ShareFolder.jsx]: i => index

File: frontend_lib/src/component/ShareDownload/ShareDownloadManagement.jsx [hunk 16]
```diff
@@ -53,12 +54,12 @@ class ShareDownloadManagement extends React.Component {
             </button>
           }
         </div>
-        {props.shareLinkList.length > 0
+        {shareLinkList.length > 0
```
Comment (basilelegal) [inline frontend_lib/src/component/ShareDownload/ShareDownloadManagement.jsx]: It's too hard to understand, this needs to be split

File: frontend/src/component/GuestPage/DownloadForm.jsx [hunk 17]
```diff
@@ -30,47 +29,49 @@ class DownloadForm extends React.Component {
...
-          <Button
-            htmlType='button'
-            bootstrapType=''
-            customClass='guestdownload__card__form__right__btn highlightBtn primaryColorBg primaryColorBgDarkenHover'
-            label={props.t('Download')}
-            onClick={props.onClickDownload}
-          />
+          <a
+            className='btn highlightBtn primaryColorBg primaryColorBgDarkenHover guestdownload__card__form__right__btn'
+            href={props.onDownloadFile()}
```
Comment (basilelegal) [inline frontend/src/component/GuestPage/DownloadForm.jsx]: getDownloadUrl

File: frontend/src/container/GuestDownload.jsx [hunk 18]
```diff
@@ -28,25 +36,43 @@ class GuestDownload extends React.Component {
...
+            isProtected: response.body.has_password
+          },
+          userName: response.body.author.public_name
         })
         break
-      default: this.sendGlobalFlashMsg(props.t('Error while loading file infos', 'warning'))
+      case 400:
+        this.sendGlobalFlashMessage(props.t('Error in the URL'))
+      default:
+        this.sendGlobalFlashMessage(props.t('Error while loading file infos'))
```
Comment (basilelegal) [inline frontend/src/container/GuestDownload.jsx]: You need to redirect to a valid url after

**Pull request merged**